src/smt/theory_str_noodler/util.cc

```diff
@@ -150,4 +150,116 @@ namespace smt::noodler::util {
+            if (!is_backtracked && (current_automaton != NUM_OF_AUTOMATA-1 || index_in_word == LENGTH_OF_WORD) && automata[current_automaton]->final.intersects_with(current_states)) {
+                // if we can accept, we save the current index, states, and word and move to the next automaton
+                STRACE("str-split-word-to-automata", tout << "Moving to next automaton\n";);
+                backtracking_indexes.push_back(index_in_word);
+                backtracking_state_sets.push_back(current_states);
+                words.push_back(current_word);
+                STRACE("str-split-word-to-automata",
+                    tout << "Current words:";
+                    for (const auto& word : words) {
+                        tout << " " << word;
+                    }
+                    tout << "\n";
+                );
+
+                ++current_automaton;
+                if (current_automaton != NUM_OF_AUTOMATA) {
+                    // index_in_word is not updated, as we stay at the same position
+                    current_states = mata::nfa::StateSet(automata[current_automaton]->initial);
+                    current_word = zstring();
+                }
+                continue;
+            }
+
+            if (index_in_word == LENGTH_OF_WORD) {
+                // we read the whole word, but we have still some automata left, we need to backtrack
+                if (current_automaton == 0) { return false; } // we cannot backtrack, i.e., the word is not accepted by the concatenation of automata
+                STRACE("str-split-word-to-automata", tout << "Backtracking at the end of the word\n";);
+
+                --current_automaton;
+                index_in_word = backtracking_indexes.back();
+                backtracking_indexes.pop_back();
+                current_states = std::move(backtracking_state_sets.back());
+                backtracking_state_sets.pop_back();
+                current_word = std::move(words.back());
+                words.pop_back();
+                is_backtracked = true;
+                continue;
+            }
+            
+            // we move by one in word and compute the new set of states
+            mata::Symbol current_symbol = word[index_in_word];
+            mata::nfa::StateSet new_current_states;
+            for (mata::nfa::State s : current_states) {
+                const mata::nfa::StatePost& transitions_from_s = automata[current_automaton]->delta[s];
+                auto transitions_from_current_symbol_it = transitions_from_s.find(current_symbol);
+                if (transitions_from_current_symbol_it != transitions_from_s.end()) {
+                    SASSERT(transitions_from_current_symbol_it->symbol == current_symbol);
+                    new_current_states = mata::nfa::StateSet::set_union(new_current_states, transitions_from_current_symbol_it->targets);
```

Comment:
Post over current symbol from the set of previous states

Reply:
Done